src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java

```diff
@@ -524,28 +531,69 @@ BlobMetadata buildAndUpload(String filePath, List<List<ChannelData>> blobData)
         crc.update(encryptedCompressedChunkData, 0, encryptedCompressedChunkDataSize);
 
         logger.logInfo(
-            "Finish building chunk in blob={}, table={}, rowCount={}, uncompressedSize={},"
-                + " compressedChunkLength={}, encryptedCompressedSize={}",
+            "Finish building chunk in blob={}, table={}, rowCount={}, startOffset={},"
+                + " uncompressedSize={}, compressedChunkLength={}, encryptedCompressedSize={},"
+                + " arrowBatchWriteMode={}",
             filePath,
             firstChannel.getFullyQualifiedTableName(),
             rowCount,
+            startOffset,
             chunkData.size(),
             compressedChunkLength,
             encryptedCompressedChunkDataSize,
-            compressedChunkLength);
+            getArrowBatchWriteMode());
       }
     }
 
     // Build blob file, and then upload to streaming ingest dedicated stage
     byte[] blob =
-        BlobBuilder.build(chunksMetadataList, chunksDataList, crc.getValue(), curDataSize);
+        BlobBuilder.build(
+            chunksMetadataList, chunksDataList, crc.getValue(), curDataSize, bdecVersion);
     if (buildContext != null) {
       buildContext.stop();
     }
 
     return upload(filePath, blob, chunksMetadataList);
   }
 
+  private Pair<byte[], Integer> compressIfNeededAndPadChunk(
```

Comment:
This function should be in `BlobBuilder`, and could you add comment for this function?